### ISSUE TITLE
xxHash instead of FNV?!

### ISSUE THREAD

**Issue by sjakobi** [haskell-unordered-containers/hashable#246]
@Mistuke mentioned this algorithm in https://github.com/haskell-unordered-containers/hashable/issues/118#issuecomment-435688921.

According to https://github.com/rurban/smhasher and https://github.com/Cyan4973/xxHash it compares nicely to FNV.

AFAIK GHC already uses this algorithm in its linker.

Has there already been any research towards using it in `hashable`?

**Comment by sjakobi** [comment]
One particular requirement of `hashable` is that lazy `ByteString` values and lazy `Text` values are hashed to the same output regardless of their internal chunking:

https://github.com/haskell-unordered-containers/hashable/blob/b8a718c73916ebfd8d05e1a9513c271e1ec4933f/src/Data/Hashable/Class.hs#L710-L717

https://github.com/haskell-unordered-containers/hashable/blob/b8a718c73916ebfd8d05e1a9513c271e1ec4933f/src/Data/Hashable/Class.hs#L730-L736

I wonder whether this property can be upheld when xxHash is used.

**Comment by Bodigrim** [comment]
> I wonder whether this property can be upheld when xxHash is used.

This is not a problem, xxHash supports stream hashing.

I did a quick experiment using https://hackage.haskell.org/package/xxhash-ffi and numbers are almost too good to be true:

```
All
  hash
    ByteString
      strict
        5:        OK
          9.91 ns ± 930 ps,       same as baseline
        8:        OK
          9.64 ns ± 834 ps,       same as baseline
        11:       OK
          11.0 ns ± 806 ps,       same as baseline
        40:       OK
          13.7 ns ± 826 ps, 54% less than baseline
        128:      OK
          17.0 ns ± 1.6 ns, 84% less than baseline
        512:      OK
          39.4 ns ± 1.8 ns, 92% less than baseline
        2^20:     OK
          63.0 μs ± 3.6 μs, 94% less than baseline
```

@phadej would you be open to using xxHash instead of FNV for `instance Hashable ByteString`? Or is depending on `xxhash.c` a no-go for `hashable`?

**Comment by phadej** [comment]
@Bodigrim it's an option to use `xxhash` for just `hash` of `ByteString` (and `Text` and `ByteArray`), but this issue is also talking about replacing the general hashing (i.e. FNV-1).

The former is easily doable; the latter isn't without breaking stuff.

**Comment by phadej** [comment]
> Or is depending on xxhash.c a no-go for hashable?

Probably GHC WASM/JS people will complain, I don't really care about these backends, so i won't.

**Comment by Bodigrim** [comment]
I'm looking just for faster hashing of `ByteArray` and friends, I'm not interested in changes to `hashInt` / `defaultHashWithSalt`.

WASM should be fine, it can cope well with bundled C libraries.

@phadej would you like me to take a stab at this?

**Comment by phadej** [comment]
> would you like me to take a stab at this?

No need. I'd end up redoing it anyway.

**Comment by Bodigrim** [comment]
With regards to WASM, I got it working fine at https://github.com/haskell-haskey/xxhash-ffi, so this should not be an issue. 

Here are benchmarks after switching to XXH3:

```
All
  10B
    Data.Digest.XXHash (c, XXH3): OK
      10.1 ns ± 838 ps
    Data.Hashable (c, FNV?):      OK
      12.1 ns ± 862 ps
  1kB
    Data.Digest.XXHash (c, XXH3): OK
      71.8 ns ± 6.5 ns
    Data.Hashable (c, FNV?):      OK
      1.14 μs ± 104 ns
  4kB
    Data.Digest.XXHash (c, XXH3): OK
      290  ns ±  26 ns
    Data.Hashable (c, FNV?):      OK
      4.68 μs ± 412 ns
  1MB
    Data.Digest.XXHash (c, XXH3): OK
      175  μs ±  14 μs
    Data.Hashable (c, FNV?):      OK
      2.87 ms ± 214 μs
```

**Comment by Bodigrim** [comment]
I've released [`xxhash-ffi-0.3`](https://hackage.haskell.org/package/xxhash-ffi-0.3) as a proof of concept. Potentially raw bindings in [`Data.Digest.XXHash.FFI.C`](https://hackage.haskell.org/package/xxhash-ffi-0.3/docs/Data-Digest-XXHash-FFI-C.html) could be separated into their own library, which `hashable` could depend upon, if it makes sense.
